src/main/java/org/quiltmc/loader/api/config/MetadataType.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api.config;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.quiltmc.loader.api.config.values.ValueTreeNode;
+
+/**
+ * A typed key to be used for setting and getting metadata of a {@link ValueTreeNode} or {@link Config}.
+ *
+ * <p>See also {@link TrackedValue.Builder#metadata}, {@link Config.Builder#metadata}, and {@link Config.SectionBuilder#metadata}
+ */
+public final class MetadataType<T, B extends MetadataType.Builder<T>> {
+	private final Class<T> metadataClass;
```

Comment:
May I recommend using `Type` instead of `Class<T>`? `Class<T>` doesn't support parameterized types, which makes `MetadataType<List<String>>` impossible without client-side unchecked operations.

Reply:
I chose to remove the class parameter altogether as it was not being used for anything.